Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,361 @@
+          state = .connected(connectedState)
+        } catch let acceptStreamError as RPCError {
+          return .failure(acceptStreamError)
+        } catch {
+          return .failure(RPCError(code: .unknown, message: "Unknown error: \(error)."))
+        }
+        return .success(.success)
+
+      case .unconnected(var unconnectedState):
+        let (stream, continuation) = AsyncStream<Void>.makeStream()
+        unconnectedState.pendingStreams.append(continuation)
+        state = .unconnected(unconnectedState)
+        return .success(.pending(stream))
+
+      case .closed:
+        return .failure(
+          RPCError(
+            code: .failedPrecondition,
+            message: "The client transport is closed."
+          )
+        )
+      }
+    }
+
+    let withStreamResult: WithStreamResult
+    do {
+      withStreamResult = try result.get()
+    } catch {
+      serverStream.outbound.finish(throwing: error)
+      clientStream.outbound.finish(throwing: error)
+      throw error
+    }
+
+    switch withStreamResult {
+    case .success:
+      ()
+    case .pending(let pendingStream):
+      for await _ in pendingStream {
+        // This loop will exit either when the task is cancelled or when the
+        // client connects and this stream can be opened.
+      }
+      try Task.checkCancellation()
+
+      try self.state.withLockedValue { state in
+        switch state {
+        case .unconnected:
+          fatalError("Invalid state.")
+        case .connected(var connectedState):
+          do {
+            try connectedState.serverTransport.acceptStream(serverStream)
+            connectedState.openStreams[connectedState.nextStreamID] = (clientStream, serverStream)
+            connectedState.nextStreamID += 1
+            state = .connected(connectedState)
+          } catch let acceptStreamError as RPCError {
+            serverStream.outbound.finish(throwing: acceptStreamError)
+            clientStream.outbound.finish(throwing: acceptStreamError)
+            throw acceptStreamError
+          } catch {
+            serverStream.outbound.finish(throwing: error)
+            clientStream.outbound.finish(throwing: error)
+            throw RPCError(code: .unknown, message: "Unknown error: \(error).")
+          }
+        case .closed:
+          let error = RPCError(
+            code: .failedPrecondition,
+            message: "The client transport is closed."
+          )
+          serverStream.outbound.finish(throwing: error)
+          clientStream.outbound.finish(throwing: error)
+          throw error
+        }
+      }
+    }
+
+    defer {
+      clientStream.outbound.finish()
+
+      let maybeEndContinuation = self.state.withLockedValue { state in
+        switch state {
+        case .unconnected:
+          fatalError("Invalid state")
+        case .connected(var connectedState):
+          connectedState.openStreams.removeValue(forKey: connectedState.nextStreamID - 1)
```

Comment:
There's a bug here: `nextStreamID - 1` isn't guaranteed to be the ID of the stream we opened in this call. Another thread can acquire the lock and open a stream between the stream being opened by this function and the defer being executed, meaning we can remove the wrong stream. 

The state machine should return the ID of the stream when the server accepts it.